docs/developers/API_Reference.md

```diff
@@ -0,0 +1,317 @@
+# VCMI Engine API Reference
+
+This document provides a comprehensive reference for developers working with the VCMI engine APIs.
+
+## Table of Contents
+
+- [Core Architecture](#core-architecture)
+- [Game State API](#game-state-api)
+- [Object System](#object-system)
+- [Bonus System API](#bonus-system-api)
+- [Scripting API](#scripting-api)
+- [Networking API](#networking-api)
+- [Serialization API](#serialization-api)
+- [Graphics and UI](#graphics-and-ui)
+
+## Core Architecture
+
+### Client-Server Model
+
+VCMI uses a client-server architecture where:
+- **Server** (`VCMI_server`) handles game logic, state management, and validation
+- **Client** (`VCMI_client`) handles user interface, input, and rendering
+- **Lib** (`VCMI_lib`) contains shared code used by both client and server
+
+### Key Base Classes
+
+#### CGameState
+Main class containing the complete game state.
+
+```cpp
+class CGameState
+{
+public:
+    std::unique_ptr<CMap> map;
+    std::array<PlayerState, PlayerColor::PLAYER_LIMIT_I> players;
+    BattleInfo *curB; // current battle
+    
+    // Game state queries
+    const CGHeroInstance * getHero(ObjectInstanceID objid) const;
+    const CGTownInstance * getTown(ObjectInstanceID objid) const;
+    // ... other methods
+};
+```
+
+#### CGObjectInstance
+Base class for all map objects.
+
+```cpp
+class CGObjectInstance : public IObjectInterface
+{
+public:
+    ObjectInstanceID id;
+    Obj ID; // object type
+    si32 subID; // object subtype
+    int3 pos; // position on map
+    
+    virtual void onHeroVisit(const CGHeroInstance * h) const;
+    virtual void initObj(vstd::RNG & rand);
+    // ... other virtual methods
+};
+```
+
+## Game State API
+
+### Player Management
+
+```cpp
+// Get player information
+const PlayerState * getPlayerState(PlayerColor player) const;
+
+// Check player status
+bool isPlayerActive(PlayerColor player) const;
+```
+
+### Hero Management
+
+```cpp
+// Create and manage heroes
+CGHeroInstance * createHero(HeroTypeID type, PlayerColor owner);
+
+// Hero movement and actions
+void moveHero(ObjectInstanceID hid, int3 dst, ui8 teleporting);
+void giveHeroNewArtifact(const CGHeroInstance *h, const CArtifact *art, ArtifactPosition pos);
```

Comment:
Do these methods actually exist? If yes - are they public or internal/private? And do they actually have such arguments?